adaptdl/adaptdl/torch/parallel.py

```diff
@@ -35,20 +40,29 @@ class AdaptiveDataParallel(DistributedDataParallel):
     adaptive batch sizes and checkpoint-restart elasticity. It automatically
     saves the given model, optimizer, and (optionally) LR scheduler whenever a
     checkpoint is triggered, and restores their states after restart. The
-    optimizer is automatically patched with AdaScale.
+    optimizer is automatically patched with the chosen scaling rule.
 
     Arguments:
         model (torch.nn.Module): Model to be distributed.
         optimizer (torch.optim.Optimizer): Optimizer used to update the given
-            model's parameters, will be patched using
-            :class:`adaptdl.torch.adascale.AdaScale`.
+            model's parameters, will be patched using subclass of
+            :class:`adaptdl.torch.scaling_rules.ScalingRuleBase`.
+        scaling_rule (Union[str, Type[ScalingRuleBase]]): Scaling rule used to
+            patch the given optimizer.
         lr_scheduler (torch.optim.lr_scheduler._LRScheduler): LR scheduler used
             to anneal the learning rate for the given optimizer.
         name (string): Unique name for each instance of this class, needed only
             if multiple instances exist.
     """
-    def __init__(self, model, optimizer, lr_scheduler=None,
-                 name="adaptdl-dataparallel", **kwargs):
+
+    available_scaling_rules = {cls.__name__: cls
+                               for cls in ScalingRuleBase.subclasses}
+
+    def __init__(self, model, optimizer,
+                 lr_scheduler=None, name="adaptdl-dataparallel",
+                 scaling_rule: Union[str, Type[SR]] = "AdaScale", **kwargs):
```

Comment:
One side effect it creates is that the user instantiated object of ScalingRuleBase is not usable right away --- A GradientNoiseScale instance needs to be set by AdaptiveDataParallel before any method of ScalingRuleBase can be invoked, especially the `step` function. 
Moreover, the `patch_optimizer` method can no longer be invoked in `__init__` of either GradientNoiseScale or ScalingRuleBase
To keep the current behavior, the workflow equivalently becomes:
```python
# user code
optim = torch.optim.SGD(model, lr=0.001)
scaling_rule = AdaScale() # not usable
adp = AdaptiveDataParallel(model, optim, scaling_rule=scaling_rule)
# inside AdaptiveDataParallel
self.gns = GradientNoiseScale(self, optimizer)
self.scaling_rule=scaling_rule
self.scaling_rule.gns = self.gns
self.scaling_rule.patch_optimizer()
```

Looks fine to you?

Reply:
Looks fine, just `self.scaling_rule.gns = self.gns` should probably be `self.scaling_rule.adp = self`, since `adp.require_backward_gradient_sync` is needed by the patched step function to decide whether to take a step or not.